### ISSUE TITLE
cc_import is more picky about header file extensions than cc_library

### ISSUE THREAD

**Issue by mpictor** [bazelbuild/bazel#6767]
### Description of the problem / feature request:

I'm currently using `cc_library` for prebuilt libraries, by setting `srcs=['path/to/libmylib.so']` and setting `hdrs` etc as normal. I thought I'd try `cc_import` since it sounded like it might work a bit differently WRT the cc_library requirement that the lib name _must_ begin with `lib`. Indeed, it works differently - but not in the way I'd expected.

With cc_import, I get errors about headers that cc_library is OK with - specifically, C++ headers like `vector` that do not have an extension.

```
ERROR: /path/to/BUILD.bazel:21:1: in hdrs attribute of cc_import rule @repo//:stlport: source file '@repo//:pub/dists/stlport/stlport/vector' is misplaced here (expected .h, .hh, .hpp, .ipp, .hxx or .inc).
ERROR: /path/to/BUILD.bazel:21:1: in hdrs attribute of cc_import rule @repo//:stlport: '@repo//:pub/dists/stlport/stlport/vector' does not produce any cc_import hdrs files (expected .h, .hh, .hpp, .ipp, .hxx or .inc).
...
```

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Assuming `bash` or similar:
```bash
mkdir bug-cc-import
cd bug-cc-import
touch WORKSPACE
cat >BUILD <<EOI
cc_import(
    name='imported',
    shared_library='libc.so',
    hdrs=['extensionlessheader'],
)

cc_binary(
    name='hello',
    deps=['imported'],
)
EOI

bazel build hello
```
bazel output:
```
ERROR: /path/to/BUILD:4:10: in hdrs attribute of cc_import rule //:imported: source file '//:extensionlessheader' is misplaced here (expected .h, .hh, .hpp, .ipp, .hxx or .inc)
ERROR: /path/to/BUILD:4:10: in hdrs attribute of cc_import rule //:imported: '//:extensionlessheader' does not produce any cc_import hdrs files (expected .h, .hh, .hpp, .ipp, .hxx or .inc)
ERROR: Analysis of target '//:hello' failed; build aborted: Analysis of target '//:imported' failed; build aborted
```
This example is too minimal to actually compile without significant changes, but cc_library does _not_ fail for headers like C++'s `<vector>`.

### What operating system are you running Bazel on?

Gentoo Linux 64-bit

### What's the output of `bazel info release`?

release 0.17.2

###  Have you found anything relevant by searching the web?

No

----

Possibly depends on #4570 (C++ Skylark API).

**Comment by rongjiecomputer** [comment]
> This example is too minimal to actually compile without significant changes, but cc_library does not fail for headers like C++'s <vector>.

`<vector>` is allowed because it is provided by system includes. Normal header files without file extension also won't work for `cc_library`.

**Comment by mpictor** [comment]
> 
> `<vector>` is allowed because it is provided by system includes. Normal header files without file extension also won't work for `cc_library`.

Are you saying `cc_library` has a whitelist for names that lack an extension but should still be accepted? That seems unmaintainable and arbitrary - what if I have an exotic system which has system headers whose names weren't added to this whitelist?

Or are you saying that `cc_library` accepts `<vector>` because it's on a system path? I'm using a CROSSTOOL with bundled libraries, so this one can't be.

**Comment by rongjiecomputer** [comment]
> Or are you saying that cc_library accepts <vector> because it's on a system path? I'm using a CROSSTOOL with bundled libraries, so this one can't be.

Something like that. Bazel will [query][1] compiler for system include paths and add them in [`cxx_builtin_include_directory`][0] of `CROSSTOOL`. If a header file's absolute path lives under any one of the `cxx_builtin_include_directory`, it will automatically be allowed without checking file extension or header dependency.

[0]: https://github.com/bazelbuild/bazel/blob/088e67d51a481fbc5bb6fff1fc517d56314aee1c/src/test/java/com/google/devtools/build/lib/analysis/mock/MOCK_CROSSTOOL#L38
[1]: https://github.com/bazelbuild/bazel/blob/master/tools/cpp/unix_cc_configure.bzl#L232

**Comment by mpictor** [comment]
So why would it make sense for system headers to be importable with `cc_library` but not `cc_import`?

**Comment by mpictor** [comment]
> `<vector>` is allowed because it is provided by system includes. Normal header files without file extension also won't work for `cc_library`.

I forgot to mention, `cc_library` gives a different error for my example `extensionlessheader` than `cc_import` does - the latter complains about the extension, while the former complains that the header doesn't exist. If I create the files, cc_library proceeds further and tries to link, while cc_import's error remains the same.

**Comment by limdor** [comment]
What is the status of this ticket? Is there any progress there? Are there any plans to change it?
I had the same issue when trying to provide rules for Qt.
I have precompiled Qt libraries and I am trying to create the rules around them for Windows without having them installed in the system with any other global flag.
For that I am using cc_import but the issue is the same:
Qt headers are intended to be used in the way
`#include <QVector>`
but this means that some files has to be exposed without extension.

**Comment by limdor** [comment]
After taking a deeper look to the issue and the code this is because the cc_library allows any file in the headers:
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppRuleClasses.java#L428
But the cc_import is only allowing cpp headers:
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/rules/cpp/CcImportRule.java#L80
The main question here would be, is this difference intentional and if yes why.

**Comment by limdor** [comment]
@hlopko is this an accepted bug? Would it be ok to change the behavior of cc_import to support the same includes like cc_library? If that is the case I can prepare a PR with the changes.

**Comment by ulfjack** [comment]
IMO, making it consistent sounds good.

**Comment by hlopko** [comment]
@oquenchil wdyt?

**Comment by oquenchil** [comment]
Fixed by #10882.
